File: python-app/domain/cache/domain_list_cache.py
from domain.cache.expiring_value_cache import ExpiringValueCache


class CachedDomainListMixin:
    """Mixin para cache de lista de dom?nios com m?todos para manipula??o de nomes."""

    _domain_cache_ttl_seconds = 30.0
    _domain_cache_max_items = 50

    def __init__(self):
        self._domain_cache = ExpiringValueCache[tuple[str, ...]](
            ttl_seconds=self._domain_cache_ttl_seconds
        )

    def _fetch_domain_names(self) -> list[str]:
        raise NotImplementedError

    def _get_domain_names_cached(self) -> list[str]:
        cached_names = self._domain_cache.get_or_set(self._fetch_domain_names_snapshot)
        return list(cached_names)

    def _fetch_domain_names_snapshot(self) -> tuple[str, ...]:
        names = self._fetch_domain_names()
        sanitized = {str(name).strip() for name in names if str(name).strip()}
        ordered = tuple(sorted(sanitized))
        return ordered[: self._domain_cache_max_items]

    def _clean_name(self, name: str) -> str:
        if "/" in name:
            return name.split("/")[-1]
        return name

    def _get_domain_view(self, prefix: str = "- ") -> list[str]:
        names = self._get_domain_names_cached()
        return [self._clean_name(name) for name in names]

    def clear_domain_cache(self) -> None:
        self._domain_cache.clear()

    def get_domain_cache_stats(self) -> dict[str, int | bool]:
        return self._domain_cache.get_stats()

File: angular-app/src/app/core/domain/cache/domain-list-cache.ts
import { ExpiringValueCache } from './expiring-value-cache';

export abstract class CachedDomainListMixin {
  protected readonly _domainCacheTtlSeconds: number = 30.0;
  protected readonly _domainCacheMaxItems: number = 50;

  private readonly _domainCache: ExpiringValueCache<readonly string[]> = new ExpiringValueCache<readonly string[]>(this._domainCacheTtlSeconds);
  private _pendingDomainNames: Promise<readonly string[]> | null = null;

  protected abstract _fetchDomainNames(): string[] | Promise<string[]>;

  protected async getDomainNamesCached(): Promise<string[]> {
    const cachedNames = this._domainCache.get();
    if (cachedNames !== null) {
      return Array.from(cachedNames);
    }

    if (!this._pendingDomainNames) {
      this._pendingDomainNames = this._fetchDomainNamesSnapshot()
        .then((snapshot) => {
          this._domainCache.set(snapshot);
          return snapshot;
        })
        .finally(() => {
          this._pendingDomainNames = null;
        });
    }

    return Array.from(await this._pendingDomainNames);
  }

  private async _fetchDomainNamesSnapshot(): Promise<readonly string[]> {
    const names = await this._fetchDomainNames();
    const sanitized = new Set(names.map(name => this._cleanName(name.trim())).filter(Boolean));
    const ordered = Array.from(sanitized).sort().slice(0, this._domainCacheMaxItems);
    return ordered as readonly string[];
  }

  protected _cleanName(name: string): string {
    if (name.includes('/')) {
      return (name.split('/').pop() || name).trim();
    }
    return name;
  }

  async getDomainView(prefix: string = '- '): Promise<string[]> {
    const names = await this.getDomainNamesCached();
    return names.map(name => `${prefix}${this._cleanName(name)}`);
  }

  clearDomainCache(): void {
    this._domainCache.clear();
    this._pendingDomainNames = null;
  }

  getDomainCacheStats(): {
    hits: number;
    misses: number;
    evictions: number;
    hasValue: boolean;
  } {
    return this._domainCache.getStats();
  }
}
